### PR TITLE
Switch to Honey SQL 2 and Toucan 2, part 2

### PR DESCRIPTION
**Pull request by camsaul**
Part 2 of 3.

- Part 1: #27742
- Part 2: #27743 
- Part 3: #27688

This PR mostly ports inconsequential changes like formatting tweaks, extra test assertions, or other things that ultimately should not affect anything.

### CODE REVIEW

File: .clj-kondo/config.edn [hunk 1]
```diff
@@ -79,7 +79,6 @@
     metabase.public-settings.premium-features/enable-enhancements?               {:namespaces ["metabase\\.models.*"]}
     metabase.query-processor.middleware.resolve-joins/maybe-resolve-source-table {:namespaces ["metabase.*"]}
     metabase.query-processor.util/normalize-token                                {:namespaces ["metabase.*"]}
-    metabase.util/key-by                                                         {:namespaces ["metabase(?:-enterprise)?.*"]}
```
Comment (camsaul) [inline .clj-kondo/config.edn:82]: I removed this a while back but forgot to remove the config

File: .clj-kondo/config.edn [hunk 2]
```diff
@@ -89,10 +88,7 @@
 
   :unresolved-var
   {:exclude
-   [colorize.core
-    metabase.test/with-gtaps
-    metabase.test/with-gtaps-for-user
-    metabase.test/with-user-attributes]}
+   [colorize.core]}
```
Comment (camsaul) [inline .clj-kondo/config.edn:91]: I removed the enterprise-only stuff from `metabase.test`. It was too hard to debug when it failed to load. I moved it to a new namespace called `metabase-enterprise.test` and updated things to use that.

File: .clj-kondo/config.edn [hunk 3]
```diff
@@ -642,8 +641,7 @@
    metabase.test/with-gtaps                                                     macros.metabase-enterprise.sandbox.test-util/with-gtaps
    metabase.test/with-persistence-enabled                                       macros.metabase.test.persistence/with-persistence-enabled
    metabase.test/with-temp-env-var-value                                        macros.metabase.test.util/with-temp-env-var-value
-   metabase.test/with-temporary-raw-setting-values                              macros.metabase.test.util/with-temporary-raw-setting-values
-   toucan.models/defmodel                                                       macros.toucan.models/defmodel}}
```
Comment (camsaul) [inline .clj-kondo/config.edn:646]: Since we imported the Toucan 2 Kondo config in https://github.com/metabase/metabase/pull/27742 we don't actually need our own version anymore.

File: enterprise/backend/src/metabase_enterprise/audit_app/pages/common.clj [hunk 4]
```diff
@@ -93,7 +93,7 @@
   ;; another
   (let [timezone (memoize/ttl sql-jdbc.sync/db-default-timezone :ttl/threshold (u/hours->ms 1))]
     (fn []
-      (timezone (mdb/db-type) (db/connection)))))
+      (timezone (mdb/db-type) {:datasource mdb.connection/*application-db*}))))
```
Comment (camsaul) [inline enterprise/backend/src/metabase_enterprise/audit_app/pages/common.clj:96]: `(db/connection)` is removed in Toucan 2, but this is what it resolves to anyway.

File: enterprise/backend/src/metabase_enterprise/enhancements/integrations/ldap.clj [hunk 5]
```diff
@@ -73,7 +74,7 @@
                           (ldap-group-membership-filter))]
       (assoc user-info :attributes (syncable-user-attributes result)))))
 
-(defenterprise-schema fetch-or-create-user! :- (class User)
+(defenterprise-schema fetch-or-create-user! :- (mi/InstanceOf User)
```
Comment (camsaul) [inline enterprise/backend/src/metabase_enterprise/enhancements/integrations/ldap.clj:77]: `(class User)` is `clojure.lang.Keyword` in Toucan 2, but I added `mi/InstanceOf` a while back which is compatible with either Toucan 1 or Toucan 2. I switched most things over to this already, but I guess I missed this spot.

File: enterprise/backend/src/metabase_enterprise/sandbox/api/gtap.clj [hunk 6]
```diff
@@ -20,11 +20,10 @@
     (db/select-one GroupTableAccessPolicy :group_id group_id :table_id table_id)
     (db/select GroupTableAccessPolicy {:order-by [[:id :asc]]})))
 
-#_{:clj-kondo/ignore [:deprecated-var]}
-(api/defendpoint-schema GET "/:id"
+(api/defendpoint GET "/:id"
```
Comment (camsaul) [inline enterprise/backend/src/metabase_enterprise/sandbox/api/gtap.clj:23]: Totally unrelated, but I updated a handful of endpoints that weren't using any schemas. I couldn't resist

File: enterprise/backend/src/metabase_enterprise/sandbox/api/gtap.clj [hunk 7]
```diff
@@ -20,11 +20,10 @@
...
   "Fetch GTAP by `id`"
   [id]
-  (api/check-404 (GroupTableAccessPolicy id)))
+  (api/check-404 (db/select-one GroupTableAccessPolicy :id id)))
```
Comment (camsaul) [inline enterprise/backend/src/metabase_enterprise/sandbox/api/gtap.clj:26]: Invoking a keyword with an integer will just give you `nil` in Toucan 2, so this won't work anymore going forward. Replaced with an equivalent form.

File: enterprise/backend/src/metabase_enterprise/serialization/dump.clj [hunk 8]
```diff
@@ -34,8 +35,11 @@
   (io/make-parents filename)
   (spit filename (yaml/generate-string obj :dumper-options {:flow-style :block})))
 
-(def ^:private as-file?
-  (comp (set (map type [Pulse Dashboard Metric Segment Field User])) type))
+(defn- as-file?
+  [instance]
+  (some (fn [model]
+          (mi/instance-of? model instance))
+        [Pulse Dashboard Metric Segment Field User]))
```
Comment (camsaul) [inline enterprise/backend/src/metabase_enterprise/serialization/dump.clj:42]: This was checking whether instances were of a `type` like `DashboardInstance`, which doesn't work in Toucan 2. Updated to use the `mi/instance-of?` helper function I added recently which works with either Toucan 1 or Toucan 2

File: enterprise/backend/src/metabase_enterprise/serialization/load.clj [hunk 9]
```diff
@@ -195,19 +195,19 @@
   (.getName (io/file path)))
 
 (defn- unresolved-names->string
-  ([entity]
-   (unresolved-names->string entity nil))
-  ([entity insert-id]
+  ([model]
+   (unresolved-names->string model nil))
+  ([model insert-id]
```
Comment (camsaul) [inline enterprise/backend/src/metabase_enterprise/serialization/load.clj:200]: A lot of the SerDes v2 code uses `entity` to mean the actual row/object getting serialized or deserialized, so this was pretty confusing. I changed it to `model` to make it a little clearer that it's referring to the actual Toucan model e.g. `User`

File: enterprise/backend/test/metabase_enterprise/sandbox/api/gtap_test.clj [hunk 10]
```diff
@@ -147,9 +147,9 @@
 
 (deftest update-gtap-test
   (testing "PUT /api/mt/gtap"
-    (mt/with-temp* [Table                  [{table-id :id}]
-                    PermissionsGroup       [{group-id :id}]
-                    Card                   [{card-id :id}]]
+    (mt/with-temp* [Table            [{table-id :id}]
+                    PermissionsGroup [{group-id :id}]
+                    Card             [{card-id :id}]]
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/sandbox/api/gtap_test.clj:152]: Formatting fix

File: enterprise/backend/test/metabase_enterprise/sandbox/api/gtap_test.clj [hunk 11]
```diff
@@ -216,8 +216,9 @@
                               :card_id  card-id-1))))
 
           (testing "Test that we can update a sandbox using the permission graph API"
-            (let [sandbox-id (db/select-one-field :id GroupTableAccessPolicy {:table_id table-id-1
-                                                                              :group_id group-id})
+            (let [sandbox-id (db/select-one-field :id GroupTableAccessPolicy
+                                                  :table_id table-id-1
+                                                  :group_id group-id)
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/sandbox/api/gtap_test.clj:221]: The fact that this ever worked in Toucan 1 was a pure accident, and I'm 95% sure it was unintentional, since there were no tests around it, it was undocumented, and I was totally confused to see this code in real life. Updated to an equivalent form that won't break when we switch to Toucan 2.

File: enterprise/backend/test/metabase_enterprise/sandbox/api/permissions_test.clj [hunk 12]
```diff
@@ -98,8 +100,8 @@
                 (testing "GTAP should be deleted from application DB"
                   (is (= []
                          (db/select GroupTableAccessPolicy
-                           :group_id (u/the-id &group)
-                           :table_id (mt/id :venues)))))
+                                    :group_id (u/the-id &group)
+                                    :table_id (mt/id :venues)))))
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/sandbox/api/permissions_test.clj:104]: Indentation fix

File: enterprise/backend/test/metabase_enterprise/sandbox/api/permissions_test.clj [hunk 13]
```diff
@@ -157,14 +159,14 @@
...
+               (:query (qp/compile
 
-                          {:database (mt/id)
-                           :query {:source-table (str "card__" (u/the-id card))}
-                           :type :query}))
-                "metabase_cache")))))
+                        {:database (mt/id)
+                         :query {:source-table (str "card__" (u/the-id card))}
+                         :type :query}))
+               "metabase_cache")))))
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/sandbox/api/permissions_test.clj:167]: Indentation fix

File: enterprise/backend/test/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions_test.clj [hunk 14]
```diff
@@ -374,42 +377,42 @@
 (deftest e2e-fks-test
   (mt/test-drivers (row-level-restrictions-fk-drivers)
     (mt/with-bigquery-fks #{:bigquery :bigquery-cloud-sdk}
-     (testing (str "1 - Creates a GTAP filtering question, looking for any checkins happening on or after 2014\n"
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions_test.clj:377]: fixed indentation

File: enterprise/backend/test/metabase_enterprise/sandbox/test_util.clj [hunk 15]
```diff
@@ -90,7 +90,7 @@
 (defmacro with-gtaps-for-user
   "Like `with-gtaps`, but for an arbitrary User. `test-user-name-or-user-id` can be a predefined test user e.g. `:rasta`
   or an arbitrary User ID."
-  {:style/indent 2}
+  {:style/indent :defn}
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/sandbox/test_util.clj:93]: The correct spec for this is actually `:defn`, because if you don't include the user or GTAPs on the very first line it should not affect the indentation of the `body` of the macro

File: enterprise/backend/test/metabase_enterprise/serialization/load_test.clj [hunk 16]
```diff
@@ -310,8 +310,6 @@
                                :sparksql  ; foreign-keys is not supported by this driver
                                ;; foreign-keys is not supported by the below driver even though it has joins
                                :bigquery-cloud-sdk))
-
-
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/serialization/load_test.clj:314]: Removed unneeded/bad whitespace

File: enterprise/backend/test/metabase_enterprise/serialization/test_util.clj [hunk 17]
```diff
@@ -126,14 +126,14 @@
           (when (.exists (io/file dump-dir))
             (.delete (io/file dump-dir))))))))
 
-(defmacro with-random-dump-dir {:style/indent 2} [[dump-dir-binding prefix] & body]
+(defmacro with-random-dump-dir {:style/indent 1} [[dump-dir-binding prefix] & body]
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/serialization/test_util.clj:129]: The correct indent spec for this was actually `1` since only the first arg is special.

File: enterprise/backend/test/metabase_enterprise/serialization/test_util.clj [hunk 18]
```diff
@@ -126,14 +126,14 @@
...
   `(do-with-random-dump-dir ~prefix (fn [~dump-dir-binding] ~@body)))
 
 (defmacro with-world
   "Run test in the context of a minimal Metabase instance connected to our test database."
+  {:style/indent 0}
   [& body]
-  `(with-temp-dpc [Database   [{~'db-id :id} (into {:name temp-db-name} (-> (data/id)
-                                                                            Database
+  `(with-temp-dpc [Database   [{~'db-id :id} (into {:name temp-db-name} (-> (data/db)
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/serialization/test_util.clj:136]: You cannot invoke a model in Toucan 2, so I replaced this with an equivalent form.

File: enterprise/backend/test/metabase_enterprise/serialization/test_util.clj [hunk 19]
```diff
@@ -330,8 +330,7 @@
                    PulseCard           [{~'pulsecard-root-id :id} {:pulse_id ~'pulse-id
                                                                    :card_id  ~'card-id-root}]
                    PulseCard           [{~'pulsecard-collection-id :id} {:pulse_id ~'pulse-id
-                                                                         :card_id  ~'card-id}
-                                                         :query {:source-table (str "card__" ~'card-id-root)}]
+                                                                         :card_id  ~'card-id}]
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/serialization/test_util.clj:333]: Toucan 2's version of Toucan 1 is actually a little better at validation and caught this -- this actually has 4 args instead of the usual one `[bindings]` or 2 `[bindings properties]`. In Toucan 1 those extra two args would be completely ignored. I went ahead and deleted them, and it didn't affect anything.

File: enterprise/backend/test/metabase_enterprise/serialization/upsert_test.clj [hunk 20]
```diff
@@ -84,40 +84,42 @@
...
+          (is (= (set id-cond) (-> e
+                                   (select-keys id-cond)
+                                   keys
+                                   set)))
+          (is (= (#'upsert/select-identical model (cond-> e
+                                                    ;; engine is a keyword but has to be a string for
+                                                    ;; HoneySQL to not interpret it as a col name
+                                                    (mi/instance-of? Database e) (update :engine name)))
+                 e)))))))
+
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/serialization/upsert_test.clj:108]: Abstracted this out of the test to make it a little easier to read

File: src/metabase/analytics/snowplow.clj [hunk 21]
```diff
@@ -23,10 +23,12 @@
    (org.apache.http.impl.client HttpClients)
    (org.apache.http.impl.conn PoolingHttpClientConnectionManager)))
 
+(set! *warn-on-reflection* true)
+
```
Comment (camsaul) [inline src/metabase/analytics/snowplow.clj:27]: At some point I want to enable the Kondo linter that requires this in every namespace that does Java interop. This way we catch reflection warnings right away. (It only affects the current namespace.)

File: src/metabase/analytics/stats.clj [hunk 22]
```diff
@@ -258,7 +258,7 @@
   "Get metrics based on pulses
   TODO: characterize by non-user account emails, # emails"
   []
-  (let [pulse-conditions {:left-join [Pulse [:= :pulse.id :pulse_id]], :where [:= :pulse.alert_condition nil]}]
+  (let [pulse-conditions {:left-join [:pulse [:= :pulse.id :pulse_id]], :where [:= :pulse.alert_condition nil]}]
```
Comment (camsaul) [inline src/metabase/analytics/stats.clj:261]: Honey SQL 2 doesn't understand Toucan 1 models (and can't be made to); Toucan 2 models are simple keywords, so this query will break when we switch over to Toucan 2. I replaced the Toucan 1 model with using the table name directly.

File: src/metabase/analytics/stats.clj [hunk 23]
```diff
@@ -279,7 +279,6 @@
      :num_alerts_per_card  (medium-histogram (vals (db-frequencies PulseCard :card_id    alert-conditions)))
      :num_cards_per_alerts (medium-histogram (vals (db-frequencies PulseCard :pulse_id   alert-conditions)))}))
 
-
```
Comment (camsaul) [inline src/metabase/analytics/stats.clj:282]: Removed extra whitespace

File: src/metabase/api/activity.clj [hunk 24]
```diff
@@ -4,13 +4,9 @@
    [clojure.string :as str]
    [compojure.core :refer [GET]]
    [medley.core :as m]
-   #_{:clj-kondo/ignore [:deprecated-var]}
-   [metabase.api.common
-    :refer [*current-user-id* defendpoint-schema define-routes]]
```
Comment (camsaul) [inline src/metabase/api/activity.clj:9]: I removed the `:refer` so we don't need to put the `ignore` comment in the `ns` declaration, which makes it hard to clean automatically.

File: src/metabase/api/dashboard.clj [hunk 25]
```diff
@@ -227,24 +227,23 @@
...
             (cond
-              (nil? (:card_id db-card)) ;; text card
-              []
+              (nil? (:card_id db-card)) ; text card
+              {}
 
               ;; cards without permissions are just a map with an :id from [[hide-unreadable-card]]
               (not (mi/model card))
-              [nil (into [card] series)]
+              {:retain nil, :discard (into [card] series)}
```
Comment (camsaul) [inline src/metabase/api/dashboard.clj:236]: Changed this to use maps instead of tuples to make it a little easier to read/debug

File: src/metabase/api/database.clj [hunk 26]
```diff
@@ -87,10 +86,10 @@
                                     :write
                                     :none))))
 
-(defn- card-database-supports-nested-queries? [{{database-id :database} :dataset_query, :as _card}]
+(defn- card-database-supports-nested-queries? [{{database-id :database, :as database} :dataset_query, :as _card}]
   (when database-id
     (when-let [driver (driver.u/database->driver database-id)]
-      (driver/supports? driver :nested-queries))))
+      (driver/database-supports? driver :nested-queries database))))
```
Comment (camsaul) [inline src/metabase/api/database.clj:92]: Unrelated change. Updated to use `database-supports?` since `supports?` is deprecated

File: src/metabase/api/database.clj [hunk 27]
```diff
@@ -596,15 +595,15 @@
    substring (s/maybe su/NonBlankString)}
   (api/read-check Database id)
   (when (and (str/blank? prefix) (str/blank? substring))
-    (throw (ex-info "Must include prefix or search" {:status-code 400})))
+    (throw (ex-info (tru "Must include prefix or search") {:status-code 400})))
```
Comment (camsaul) [inline src/metabase/api/database.clj:598]: Fixed missing i18n

File: src/metabase/api/embed.clj [hunk 28]
```diff
@@ -237,10 +237,10 @@
 (defn card-for-unsigned-token
   "Return the info needed for embedding about Card specified in `token`. Additional `constraints` can be passed to the
   `public-card` function that fetches the Card."
-  {:style/indent 1}
   [unsigned-token & {:keys [embedding-params constraints]}]
-  (let [card-id        (embed/get-in-unsigned-token-or-throw unsigned-token [:resource :question])
-        token-params   (embed/get-in-unsigned-token-or-throw unsigned-token [:params])]
+  {:pre [((some-fn empty? sequential?) constraints) (even? (count constraints))]}
```
Comment (camsaul) [inline src/metabase/api/embed.clj:241]: This was being called like

```clj
(card-for-unsigned-token <token> :constraints [[:x 1]])
```

which was passed to Toucan as something like

```clj
(db/select Card :token <token> [:x 1])
```

This was never supposed to work, but somehow it did in Toucan 1. In Toucan 2, I don't plan on making this work, since the original behavior was undocumented an unintentional. I changed things to ensure we execute something like

```clj
(db/select Card :token <token> [x 1)
```

one instead which is the equivalent Toucan 2-compatibile version.

File: src/metabase/cmd/rotate_encryption_key.clj [hunk 29]
```diff
@@ -35,7 +35,7 @@
     (jdbc/with-db-transaction [t-conn {:datasource (mdb.connection/data-source)}]
       (doseq [[id details] (db/select-id->field :details Database)]
         (when (encryption/possibly-encrypted-string? details)
-          (throw (ex-info (trs "Can't decrypt app db with MB_ENCRYPTION_SECRET_KEY") {:database-id id})))
+          (throw (ex-info (trs "Can''t decrypt app db with MB_ENCRYPTION_SECRET_KEY") {:database-id id})))
```
Comment (camsaul) [inline src/metabase/cmd/rotate_encryption_key.clj:38]: Fixed busted i18n string

File: src/metabase/driver/sql/query_processor.clj [hunk 30]
```diff
@@ -863,15 +862,15 @@
   [:= (->honeysql driver field) (->honeysql driver value)])
 
 (defn- correct-null-behaviour
-  [driver [op & args]]
-  (let [field-arg (mbql.u/match-one args
-                    FieldInstance &match
```
Comment (camsaul) [inline src/metabase/driver/sql/query_processor.clj:868]: The SQL QP never actually sees `FieldInstances` anymore so this was dead code

File: src/metabase/models/card.clj [hunk 31]
```diff
@@ -93,12 +94,17 @@
...
+    ;; mega HACK FIXME -- don't update this stuff when doing deserialization because it might differ from what's in the
+    ;; YAML file and break tests like [[metabase-enterprise.serialization.v2.e2e.yaml-test/e2e-storage-ingestion-test]].
+    ;; The root cause of this issue is that we're generating Cards that have a different Database ID or Table ID from
+    ;; what's actually in their query -- we need to fix [[metabase.test.generate]], but I'm not sure how to do that
+    (not mi/*deserializing?*)
+    (merge (when-let [{:keys [database-id table-id]} (and query-type
+                                                          (query/query->database-and-table-ids outer-query))]
+             {:database_id database-id
+              :table_id    table-id
+              :query_type  (keyword query-type)}))))
```
Comment (camsaul) [inline src/metabase/models/card.clj:107]: Toucan 2 is optimizes out things that haven't changed (e.g. the `database_id`) before calling `before-update` functions like this, which means `(merge {:database_id ...} card)` actually breaks things in Toucan 2.

This change shouldn't affect anything here since this PR still uses Toucan 1, but we can merge it in now to make sure and then not need to change it in the next PR.

File: src/metabase/models/card.clj [hunk 32]
```diff
@@ -410,7 +417,7 @@
         snippets      (some->> card :dataset_query :native :template-tags vals (keep :snippet-id))]
     (set/union
       (when (and (string? source-table)
-                 (.startsWith ^String source-table "card__"))
+                 (str/starts-with? source-table "card__"))
```
Comment (camsaul) [inline src/metabase/models/card.clj:420]: https://guide.clojure.style/#strings

File: src/metabase/models/login_history.clj [hunk 33]
```diff
@@ -86,10 +87,8 @@
...
-  (merge
-   models/IModelDefaults
-   {:post-select post-select
-    :post-insert post-insert
-    :pre-update  pre-update}))
+(mi/define-methods
+ LoginHistory
+ {:post-select post-select
+  :post-insert post-insert
+  :pre-update  pre-update})
```
Comment (camsaul) [inline src/metabase/models/login_history.clj:94]: I added this helper that works (or will work, at least) with Toucan 2 recently and updated everything to use it, but I guess I forgot to update this model.

File: src/metabase/models/permissions.clj [hunk 34]
```diff
@@ -573,7 +573,7 @@
   (set-has-full-permissions? permissions-set (application-perms-path perm-type)))
 
 (s/defn perms-objects-set-for-parent-collection :- #{Path}
-  "Implementation of `IModel` `perms-objects-set` for models with a `collection_id`, such as Card, Dashboard, or Pulse.
+  "Implementation of `perms-objects-set` for models with a `collection_id`, such as Card, Dashboard, or Pulse.
```
Comment (camsaul) [inline src/metabase/models/permissions.clj:576]: `perms-objects-set` was never part of `IModel` and especially isn't now since it's a multimethod

File: src/metabase/models/pulse.clj [hunk 35]
```diff
@@ -229,7 +229,7 @@
   "Hydrate Pulse or Alert with the Fields needed for sending it."
   [notification :- (mi/InstanceOf Pulse)]
   (-> notification
-      (hydrate :creator :cards :dashboard [:channels :recipients])
+      (hydrate :creator :cards [:channels :recipients])
```
Comment (camsaul) [inline src/metabase/models/pulse.clj:232]: There is no hydration method for `:dashboard`, in Toucan 2 this throws an error so I removed the key. Toucan 1 just silently ignores undefined hydration keys 😢

File: src/metabase/models/revision.clj [hunk 36]
```diff
@@ -51,7 +51,7 @@
 (defmethod diff-str :default
   [model o1 o2]
   (when-let [[before after] (data/diff o1 o2)]
-    (diff-string (:name model) before after)))
+    (diff-string (name model) before after)))
```
Comment (camsaul) [inline src/metabase/models/revision.clj:54]: Toucan 1 models have a `:name` key that is basically the same as calling `name` on them. In Toucan 2 since models are keywords `name` gives you the same answer while `:name` does not work.

```clj
;;; Toucan 1
Card
#metabase.models.card.CardInstance{:table :report_card, :name "Card", ...}

(:name Card)
;; => "Card"

(name Card)
;; => "Card")

;;; Toucan 2
Card
:metabase.models.card/Card

(:name Card)
;; => nil

(name Card)
;; => "Card")
```

File: src/metabase/models/serialization/base.clj [hunk 37]
```diff
@@ -81,7 +82,8 @@
   - The primary key might still be attached, during extraction.
   - `:label` is optional
   - The logic to guess the leaf part of the path is in [[infer-self-path]], for use in overriding."
-  (fn [model _] model))
+  {:arglists '([model-name instance])}
```
Comment (camsaul) [inline src/metabase/models/serialization/base.clj:85]: I added `:arglists` metadata to a bunch of SerDes things

File: src/metabase/models/setting/cache.clj [hunk 38]
```diff
@@ -85,7 +85,7 @@
           (db/simple-insert! 'Setting :key settings-last-updated-key, :value current-timestamp-as-string-honeysql)
           (catch java.sql.SQLException e
             ;; go ahead and log the Exception anyway on the off chance that it *wasn't* just a race condition issue
-            (log/error (trs "Error inserting a new Setting: {0}"
+            (log/error (trs "Error updating Settings last updated value: {0}"
```
Comment (camsaul) [inline src/metabase/models/setting/cache.clj:88]: Better error message

File: src/metabase/query_processor/middleware/add_dimension_projections.clj [hunk 39]
```diff
@@ -85,7 +85,11 @@
     ;;
     ;; Not sure this isn't broken. Probably better to have [[metabase.query-processor.util.add-alias-info]] do the name
     ;; deduplication instead.
-    (let [unique-name (comp (mbql.u/unique-name-generator) :name Field)]
+    (let [name-generator (mbql.u/unique-name-generator)
+          unique-name    (fn [field-id]
+                           (qp.store/fetch-and-store-fields! #{field-id})
+                           (let [field (qp.store/field field-id)]
+                             (name-generator (:name field))))]
```
Comment (camsaul) [inline src/metabase/query_processor/middleware/add_dimension_projections.clj:92]: Invoking a model as a function doesn't work in Toucan 2

File: src/metabase/server/middleware/session.clj [hunk 40]
```diff
@@ -232,21 +231,20 @@
    (fn [db-type max-age-minutes session-type enable-advanced-permissions?]
      (first
       (db/honeysql->sql
-       (cond->
-         {:select    [[:session.user_id :metabase-user-id]
-                      [:user.is_superuser :is-superuser?]
-                      [:user.locale :user-locale]]
-          :from      [[Session :session]]
```
Comment (camsaul) [inline src/metabase/server/middleware/session.clj:239]: Using a model in a Honey SQL query doesn't work in Honey SQL 2 so this change will set us up for the switch

File: test/metabase/api/dashboard_test.clj [hunk 41]
```diff
@@ -198,9 +198,9 @@
               (mt/user-http-request :rasta :post 200 "dashboard" {:name                dashboard-name
                                                                   :collection_id       (u/the-id collection)
                                                                   :collection_position 1000})
-              (is (= #metabase.models.dashboard.DashboardInstance{:collection_id true, :collection_position 1000}
```
Comment (camsaul) [inline test/metabase/api/dashboard_test.clj:201]: No such thing as a `DashboardInstance` in Toucan 2

File: test/metabase/api/database_test.clj [hunk 42]
```diff
@@ -772,25 +772,28 @@
...
           (is (not= expected
                     (into {} (db/select-one [Database :cache_field_values_schedule :metadata_sync_schedule] :id (u/the-id db)))))))
       (testing "We can if we mark `:let-user-control-scheduling`"
         (mt/with-temp Database [db {:engine "h2", :details (:details (mt/db))}]
-          (mt/user-http-request :crowberto :put 200 (format "database/%d" (u/the-id db))
-                                (-> db
-                                    (assoc :schedules attempted)
-                                    (assoc-in [:details :let-user-control-scheduling] true)))
+          (is (=? expected
+                  (mt/user-http-request :crowberto :put 200 (format "database/%d" (u/the-id db))
```
Comment (camsaul) [inline test/metabase/api/database_test.clj:783]: More test assertions

File: test/metabase/models/metric_test.clj [hunk 43]
```diff
@@ -26,13 +26,13 @@
     (mt/with-temp Metric [{:keys [id]} {:creator_id (mt/user->id :rasta)}]
       (testing "you should not be able to change the creator_id of a Metric"
         (is (thrown-with-msg?
-             UnsupportedOperationException
+             Exception
```
Comment (camsaul) [inline test/metabase/models/metric_test.clj:29]: These get wrapped in `ExceptionInfo` in Toucan 2

File: src/metabase/server/middleware/session.clj (hunk 40 above)
Comment (bshepherdson) [reply src/metabase/server/middleware/session.clj:239]: Is there a Toucan 2-friendly way to go from the Toucan 2 model keyword to the actual table? I'm imagining I'm familiar with the model keyword (`:dashboard` or `:card`) and don't want to have to remember the legacy table names (`:report_dashboard` and `:report_card`).

Then again, if I'm using HoneySQL directly, we're already deep into the table structure.

**Review by bshepherdson — APPROVED** [review]

**Review by metamben — APPROVED** [review]
That's a lot of unrelated changes. But I like them!

Is the plan that some of the changes that are needed for compatibility (e.g., `Exception` instead of `ExceptionInfo`) will be streamlined after the arrival of toucan2?

**Comment by camsaul** [comment]
@bshepherdson mentioned in Slack but in case you didn't see it: Toucan 2 does have `t2/table-name`, so you could do

```clj
(t2/table-name Pulse) ;; => :pulse
```

and use that in the Honey SQL queries. But since we don't have access to `t2/table-name` (yet) I didn't think using `:pulse` directly was that bad for the time being and writing a helper version of it wasn't really necessary. Based on the Slack convo, it seems like other people do like getting the table name from the model so we can use that more in the future. I might turn around and update all these things I changed here to use it if the consensus is that's the preferable way of doing things

File: src/metabase/api/card.clj [hunk 44]
```diff
@@ -181,7 +181,6 @@
   (let [raw-card (db/select-one Card :id id)
         card (-> raw-card
                  (hydrate :creator
-                          :bookmarked
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/api/card.clj:184]: Sorry, I forgot to call this out. There is actually no hydration method for `:bookmarked` anymore. So I removed this because it's a no-op in Toucan 1. In Toucan 2, it's an error

**Comment by camsaul** [comment]
@metamben RE

> Is the plan that some of the changes that are needed for compatibility (e.g., Exception instead of ExceptionInfo) will be streamlined after the arrival of toucan2?

I'm not 100% what you mean by that, but if you mean that we will be able to use `ExceptionInfo` more often inside `thrown-with-msg?` or `thrown?` in the future then the answer will definitely be yet. Toucan 2 catches all exceptions and adds extra context so any error from it will come back as `ExceptionInfo`.

**Comment by metamben** [comment]
> we will be able to use `ExceptionInfo` more often inside `thrown-with-msg?` or `thrown?` in the future

Yes, this is an example where I think the code should be changed after the migration. There are others, like using the model name keyword (e.g., `:user`) instead of the table name keyword (`:core_user`), if I understand correctly.
